### PR TITLE
[instCombine][bugfix] Fix crash caused by using of cast in instCombineSVECmpNE

### PR DESCRIPTION
**Pull request by cceerczw**
Func instCombineSVECmpNE is used to identify specific pattern of instruction 'svecmene',
and then predict its result, use the result to replace instruction 'svecmene'. The specific pattern can be descriped below:
1.The svecmpne must compare all elements of vec.
2.The svecmpne inst compare its ves with zero.
3.The vec in svecmpne inst is generated by inst dupqlane, and the copy value of this dupqlane must be zero.
In NO.3 above, func instCombineSVECmpNE uses 'cast' to transform op1 of dupqlane without checking if the cast is success, then generate a crash in some situation.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 1]
```diff
@@ -1174,7 +1174,8 @@ static std::optional<Instruction *> instCombineSVECmpNE(InstCombiner &IC,
     return std::nullopt;
 
   // Where the dupq is a lane 0 replicate of a vector insert
-  if (!cast<ConstantInt>(DupQLane->getArgOperand(1))->isZero())
+  auto *DupQLaneOp1 = dyn_cast<ConstantInt>(DupQLane->getArgOperand(1));
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:1177]: Up to you but I think `DupQLaneIdx` would be more readable.

**Review by paulwalker-arm — COMMENTED** [review]
Fix looks good but the test case can be significantly reduced.  It looks like you can just  add to `llvm/test/Transforms/InstCombine/AArch64/sve-intrinsic-opts-cmpne.ll` by cloning `dupq_b_0` as `dupq_b_idx` whereby it takes an `%idx` function parameter that it uses when calling `sve.dupq.lane`.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 1 above)
Comment (wuxie2022) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:1177]: Ok, fixed.

**Comment by wuxie2022** [comment]
> Fix looks good but the test case can be significantly reduced. It looks like you can just add to `llvm/test/Transforms/InstCombine/AArch64/sve-intrinsic-opts-cmpne.ll` by cloning `dupq_b_0` as `dupq_b_idx` whereby it takes an `%idx` function parameter that it uses when calling `sve.dupq.lane`.

Thank you for your review! I have simplified the case in 'sve-inst-combine-cmpne.ll'. But it can't be merged into sve-intrinsic-opts-cmpne.ll, because the compile command of my new test case must contain '-mtriple=aarch64-unknown-linux-gnu' and '-O2'.

**Comment by paulwalker-arm** [comment]
> > Fix looks good but the test case can be significantly reduced. It looks like you can just add to `llvm/test/Transforms/InstCombine/AArch64/sve-intrinsic-opts-cmpne.ll` by cloning `dupq_b_0` as `dupq_b_idx` whereby it takes an `%idx` function parameter that it uses when calling `sve.dupq.lane`.
> 
> Thank you for your review! I have simplified the case in 'sve-inst-combine-cmpne.ll'. But it can't be merged into sve-intrinsic-opts-cmpne.ll, because the compile command of my new test case must contain '-mtriple=aarch64-unknown-linux-gnu' and '-O2'.

Are you sure?  `sve-intrinsic-opts-cmpne.ll` already contains a `target triple` line and I think the only reason you need `-O2` is because you're relying on function inlining?  I've just tried my suggestion on the current main branch and adding the following to `sve-intrinsic-opts-cmpne.ll` triggers the original failure for me. Do you not see the same?

```
define <vscale x 16 x i1> @dupq_b_idx(i64 %idx) #0 {                                
  %1 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.ptrue.nxv16i1(i32 31)         
  %2 = tail call <vscale x 16 x i8> @llvm.vector.insert.nxv16i8.v16i8(<vscale x 16 x i8> undef,
    <16 x i8> <i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0,                      
               i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0>, i64 0)              
  %3 = tail call <vscale x 16 x i8> @llvm.aarch64.sve.dupq.lane.nxv16i8(<vscale x 16 x i8> %2 , i64 %idx)
  %4 = tail call <vscale x 2 x i64> @llvm.aarch64.sve.dup.x.nxv2i64(i64 0)          
  %5 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.cmpne.wide.nxv16i8(<vscale x 16 x i1> %1, <vscale x 16 x i8> %3, <vscale x 2 x i64> %4)
  ret <vscale x 16 x i1> %5                                                         
} 
```

**Comment by wuxie2022** [comment]
> > > Fix looks good but the test case can be significantly reduced. It looks like you can just add to `llvm/test/Transforms/InstCombine/AArch64/sve-intrinsic-opts-cmpne.ll` by cloning `dupq_b_0` as `dupq_b_idx` whereby it takes an `%idx` function parameter that it uses when calling `sve.dupq.lane`.
> > 
> > 
> > Thank you for your review! I have simplified the case in 'sve-inst-combine-cmpne.ll'. But it can't be merged into sve-intrinsic-opts-cmpne.ll, because the compile command of my new test case must contain '-mtriple=aarch64-unknown-linux-gnu' and '-O2'.
> 
> Are you sure? `sve-intrinsic-opts-cmpne.ll` already contains a `target triple` line and I think the only reason you need `-O2` is because you're relying on function inlining? I've just tried my suggestion on the current main branch and adding the following to `sve-intrinsic-opts-cmpne.ll` triggers the original failure for me. Do you not see the same?
> 
> ```
> define <vscale x 16 x i1> @dupq_b_idx(i64 %idx) #0 {                                
>   %1 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.ptrue.nxv16i1(i32 31)         
>   %2 = tail call <vscale x 16 x i8> @llvm.vector.insert.nxv16i8.v16i8(<vscale x 16 x i8> undef,
>     <16 x i8> <i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0,                      
>                i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0>, i64 0)              
>   %3 = tail call <vscale x 16 x i8> @llvm.aarch64.sve.dupq.lane.nxv16i8(<vscale x 16 x i8> %2 , i64 %idx)
>   %4 = tail call <vscale x 2 x i64> @llvm.aarch64.sve.dup.x.nxv2i64(i64 0)          
>   %5 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.cmpne.wide.nxv16i8(<vscale x 16 x i1> %1, <vscale x 16 x i8> %3, <vscale x 2 x i64> %4)
>   ret <vscale x 16 x i1> %5                                                         
> } 
> ```



> > > Fix looks good but the test case can be significantly reduced. It looks like you can just add to `llvm/test/Transforms/InstCombine/AArch64/sve-intrinsic-opts-cmpne.ll` by cloning `dupq_b_0` as `dupq_b_idx` whereby it takes an `%idx` function parameter that it uses when calling `sve.dupq.lane`.
> > 
> > 
> > Thank you for your review! I have simplified the case in 'sve-inst-combine-cmpne.ll'. But it can't be merged into sve-intrinsic-opts-cmpne.ll, because the compile command of my new test case must contain '-mtriple=aarch64-unknown-linux-gnu' and '-O2'.
> 
> Are you sure? `sve-intrinsic-opts-cmpne.ll` already contains a `target triple` line and I think the only reason you need `-O2` is because you're relying on function inlining? I've just tried my suggestion on the current main branch and adding the following to `sve-intrinsic-opts-cmpne.ll` triggers the original failure for me. Do you not see the same?
> 
> ```
> define <vscale x 16 x i1> @dupq_b_idx(i64 %idx) #0 {                                
>   %1 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.ptrue.nxv16i1(i32 31)         
>   %2 = tail call <vscale x 16 x i8> @llvm.vector.insert.nxv16i8.v16i8(<vscale x 16 x i8> undef,
>     <16 x i8> <i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0,                      
>                i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0>, i64 0)              
>   %3 = tail call <vscale x 16 x i8> @llvm.aarch64.sve.dupq.lane.nxv16i8(<vscale x 16 x i8> %2 , i64 %idx)
>   %4 = tail call <vscale x 2 x i64> @llvm.aarch64.sve.dup.x.nxv2i64(i64 0)          
>   %5 = tail call <vscale x 16 x i1> @llvm.aarch64.sve.cmpne.wide.nxv16i8(<vscale x 16 x i1> %1, <vscale x 16 x i8> %3, <vscale x 2 x i64> %4)
>   ret <vscale x 16 x i1> %5                                                         
> } 
> ```

When I try your suggestion, the compiler didn't generate any failure. I have also test in https://www.godbolt.org/ using 'opt(trunk)'. Besides, the original failure which I want to fix is like:

**Comment by paulwalker-arm** [comment]
Sorry I really don't mean to be a pain here and I'll accept this version of the PR if you continue to see something different to me, but is it possible you're not using a debug or asserts build?

https://www.godbolt.org/z/8McfPMqTz shows the same error quite clearly.

**Comment by wuxie2022** [comment]
> Sorry I really don't mean to be a pain here and I'll accept this version of the PR if you continue to see something different to me, but is it possible you're not using a debug or asserts build?

The result of https://www.godbolt.org/z/8McfPMqTz shows that the recommended case crashes when compiled with the debug version of the compiler, but it does not crash when compiled with the release version. On the other hand, the case I uploaded crashes even when compiled with the release version of the compiler. So what I don't understand is, if this case can only detect the crash error when compiled with the debug version of the compiler, is this case really meaningful? If I receive your confirmation, I will make the necessary changes. Once again, thank you for your review!

**Comment by paulwalker-arm** [comment]
Yes, having a test case that only exposes a failure when asserts are enable is legitimate and expected because all active LLVM development should be done when asserts are enabled, otherwise you're minimising your test coverage.

Whilst the other test case crashes with a release compiler today, a crash is inherently unexplained and so the same test might not crash in the future. For example, you already mention the need for `-O2`, which enables many code paths that are unrelated to the code you've fixed.  The simpler test makes it clearer how the code is linked to the affected assert.

**Comment by wuxie2022** [comment]
> Yes, having a test case that only exposes a failure when asserts are enable is legitimate and expected because all active LLVM development should be done when asserts are enabled, otherwise you're minimising your test coverage.
> 
> Whilst the other test case crashes with a release compiler today, a crash is inherently unexplained and so the same test might not crash in the future. For example, you already mention the need for `-O2`, which enables many code paths that are unrelated to the code you've fixed. The simpler test makes it clearer how the code is linked to the affected assert.

Thank you for your explanation. I will modify the case.

**Review by paulwalker-arm — APPROVED** [review]

**Comment by cceerczw** [comment]
Hi,@paulwalker-arm, thanks for the review again, can this patch merge?

**Comment by paulwalker-arm** [comment]
> Hi,@paulwalker-arm, thanks for the review again, can this patch merge?

I see no reason why not.  The windows failure looks unrelated to this work.

**Comment by cceerczw** [comment]
> > Hi,@paulwalker-arm, thanks for the review again, can this patch merge?
> 
> I see no reason why not. The windows failure looks unrelated to this work.
yes, so can you help me to merge this patch?

**Comment by cceerczw** [comment]
> yes, so can you help me to merge this patch?

yes, so can you help me to merge this patch?

**Comment by cceerczw** [comment]
> > Hi,@paulwalker-arm, thanks for the review again, can this patch merge?
> 
> I see no reason why not. The windows failure looks unrelated to this work.

yes, so can you help me to merge this patch?

**Pull request merged**